### DISCUSSION TITLE
Question: Can I use this for Intercom Custom Domain?

### DISCUSSION THREAD

**Discussion by mikeesouth** [dotnet/yarp#681] [Q&A] [answered]
Hi,

I'm sorry if this is not the right place to ask this question but I do think that others will search for this use case and that an answer may help more people than only myself.

I'm using a third party product on my site: Intercom. It's a live chat messenger that you can embed on your webpage. It supports help articles to be hosted at a custom domain: https://intercom.help/your-company (by default). They also support custom domains so that you can show the help portal on help.your-company.com but that requires a reverse proxy in order for SSL to work. They supply a guide for custom SSL using a reverse proxy configured in Apache and nginx but not for anything in .NET Core. 

I thought YARP would help me host this in an Azure Web App running .NET 5. I followed the YARP getting started guide but I only get 404 on my site and I have no clue on where to start troubleshooting this.

The nginx example is like this:
```
server {
  listen 443 ssl;
  server_name your-help-site.custom-domain.com; # replace this with your domain

  ssl_certificate /path/to/your/fullchain.pem;
  ssl_certificate_key /path/to/your/privatekey.pem;

  location / {
    # using "set" is important as IP addresses of Intercom servers 
    # changes dynamically. "set" enables nginx to follow dynamic IPs
    set $intercom "https://custom.intercom.help:443"; 
    proxy_set_header Host $host;
    proxy_pass $intercom;
  }
```

I have this code in Startup.cs of my .NET 5 application:
```c#
public class Startup
{
    public IConfiguration Configuration { get; }

    public Startup(IConfiguration configuration)
    {
        Configuration = configuration;
    }

    // This method gets called by the runtime. Use this method to add services to the container.
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddReverseProxy().LoadFromConfig(Configuration.GetSection("ReverseProxy"));
    }

    // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        if (env.IsDevelopment())
        {
            app.UseDeveloperExceptionPage();
        }

        app.UseRouting();
        app.UseEndpoints(endpoints =>
        {
            endpoints.MapReverseProxy();
        });
    }
}
```

And I have this ReverseProxy section in my appsettings.json:
```json
"ReverseProxy": {
    "Routes": [
      {
        "RouteId": "route1",
        "ClusterId": "cluster1",
        "Match": {
          "Path": "{**catch-all}"
        }
      }
    ],
    "Clusters": {
      "cluster1": {
        "Destinations": {
          "cluster1/destination1": {
            "Address": "https://custom.intercom.help/"
          }
        }
      }
    }
  }
```

Could this work or am I completely missing the point here? :)
I have an SSL certifcate on my Azure Web App and I have it running the .NET 5 runtime and I do have "<TargetFramework>netcoreapp5.0</TargetFramework>" in my .csproj.

Any help is greatly appreciated!

**Comment by samsp-msft** [comment, marked as answer]
Converting to a discussion, as this is a troubleshooting question.

Do you have the output from the console (or logging if hosted) from the proxy when the server is started up, and the request is made to it.
You might want to stick
```
   "Logging": {
    "LogLevel": {
      "Default": "Information"
    }
  },
```
At the top of your config file to get additional log messages. This will then tell you more about what the server is up to, and details for each request.

**Reply by mikeesouth** [reply]
@samsp-msft thanks for your tip - the log files clearly indicated that I needed to set a host value in my configuration. I've been testing alot and setting the host value in the route matcher helped. But I also used YARP version "1.0.0-preview.1.*" and after upgrading to "1.0.0-preview.8.*" it is no longer needed. So I just upgraded to "1.0.0-preview.8.*" and it works - kinda.

I can route to https://www.google.com/ and I do see the google.com website - so far so good. But if I route to https://custom.intercom.help/ I get a 404. That's probably expected because if I navigate to that URL in my browser I also get a 404. I think they're using X-Forwarded-Host or something to show the correct page for our custom domain.

They have instructions on how to setup the reverse proxy here: https://developers.intercom.com/installing-intercom/docs/set-up-your-custom-domain#section-using-your-own-ssl-certificate-with-apache-or-nginx

But the instructions are for apache and nginx. Do you have any idea on how I replicate these settings in YARP and/or other features in .NET 5?

I'm thinking that the key is the Apache call to `ProxyPreserveHost On` and the nginx call to `proxy_set_header Host $host;` (host is help.your-company.com). Can I set an equivalent option in YARP?

**Reply by mikeesouth** [reply]
Ok, a little more progress but still not working. When calling from my production host with the route transforms specified below, I actually get a 502 error (Bad Gateway). I guess some transforms are missing/are incorrect.
```
{ "RequestHeaderOriginalHost": "true" },
```

I also tried these settings, still 502:
```
{ "ClientCert": "X-Client-Cert" },
{ "RequestHeaderOriginalHost": "true" },
{ "X-Forwarded": "" }
```

**Reply by Tratcher** [reply]
What shows up in the proxy logs when you get the 502? And the destination server logs?

**Reply by mikeesouth** [reply]
> What shows up in the proxy logs when you get the 502? And the destination server logs?

Unfortunately, I can't see the logs on the destination server and I cannot know more what they want me to provide than the "custom domain" help article I linked earlier ().

I cant see any errors in the log but I get an IIS HTTP Error page that says this:
```
HTTP Error 502.0 - Bad Gateway
The specified CGI application encountered an error and the server terminated the process.

Most likely causes:
The CGI application did not return a valid set of HTTP errors.
A server acting as a proxy or gateway was unable to process the request due to an error in a parent gateway.
```

The full HTML of the error page is in the logs:
```
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>IIS Detailed Error - 502.0 - Bad Gateway</title>
<style type="text/css"><!--body{margin:0;font-size:.7em;font-family:Verdana,Arial,Helvetica,sans-serif;}code{margin:0;color:#006600;font-size:1.1em;font-weight:bold;}.config_source code{font-size:.8em;color:#000000;}pre{margin:0;font-size:1.4em;word-wrap:break-word;}ul,ol{margin:10px 0 10px 5px;}ul.first,ol.first{margin-top:5px;}fieldset{padding:0 15px 10px 15px;word-break:break-all;}.summary-container fieldset{padding-bottom:5px;margin-top:4px;}legend.no-expand-all{padding:2px 15px 4px 10px;margin:0 0 0 -12px;}legend{color:#333333;;margin:4px 0 8px -12px;_margin-top:0px;font-weight:bold;font-size:1em;}a:link,a:visited{color:#007EFF;font-weight:bold;}a:hover{text-decoration:none;}h1{font-size:2.4em;margin:0;color:#FFF;}h2{font-size:1.7em;margin:0;color:#CC0000;}h3{font-size:1.4em;margin:10px 0 0 0;color:#CC0000;}h4{font-size:1.2em;margin:10px 0 5px 0;}#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:"trebuchet MS",Verdana,sans-serif;color:#FFF;background-color:#5C87B2;}#content{margin:0 0 0 2%;position:relative;}.summary-container,.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}.content-container p{margin:0 0 10px 0;}#details-left{width:35%;float:left;margin-right:2%;}#details-right{width:63%;float:left;overflow:hidden;}#server_version{width:96%;_height:1px;min-height:1px;margin:0 0 5px 0;padding:11px 2% 8px 2%;color:#FFFFFF;background-color:#5A7FA5;border-bottom:1px solid #C1CFDD;border-top:1px solid #4A6C8E;font-weight:normal;font-size:1em;color:#FFF;text-align:right;}#server_version p{margin:5px 0;}table{margin:4px 0 4px 0;width:100%;border:none;}td,th{vertical-align:top;padding:3px 0;text-align:left;font-weight:normal;border:none;}th{width:30%;text-align:right;padding-right:2%;font-weight:bold;}thead th{background-color:#ebebeb;width:25%;}#details-right th{width:20%;}table tr.alt td,table tr.alt th{}.highlight-code{color:#CC0000;font-weight:bold;font-style:italic;}.clear{clear:both;}.preferred{padding:0 5px 2px 5px;font-weight:normal;background:#006633;color:#FFF;font-size:.8em;}--></style>
</head><body><div id="content"><div class="content-container"><h3>HTTP Error 502.0 - Bad Gateway</h3><h4>The specified CGI application encountered an error and the server terminated the process.</h4></div><div class="content-container"><fieldset><h4>Most likely causes:</h4><ul> 	<li>The CGI application did not return a valid set of HTTP errors.</li> 	<li>A server acting as a proxy or gateway was unable to process the request due to an error in a parent gateway.</li> </ul></fieldset></div><div class="content-container"><fieldset><h4>Things you can try:</h4><ul> 	<li>Use DebugDiag to troubleshoot the CGI application.</li> 	<li>Determine if a proxy or gateway is responsible for this error.</li> </ul></fieldset></div>
<div class="content-container"><fieldset><h4>Detailed Error Information:</h4><div id="details-left"><table border="0" cellpadding="0" cellspacing="0"><tr class="alt"><th>Module</th><td>&nbsp;&nbsp;&nbsp;AspNetCoreModuleV2</td></tr><tr><th>Notification</th><td>&nbsp;&nbsp;&nbsp;ExecuteRequestHandler</td></tr><tr class="alt"><th>Handler</th><td>&nbsp;&nbsp;&nbsp;aspNetCore</td></tr><tr><th>Error Code</th><td>&nbsp;&nbsp;&nbsp;0x00000000</td></tr>
</table></div><div id="details-right"><table border="0" cellpadding="0" cellspacing="0"><tr class="alt"><th>Requested URL</th><td>&nbsp;&nbsp;&nbsp;http://localhost:80/</td></tr><tr><th>Physical Path</th><td>&nbsp;&nbsp;&nbsp;C:\home\site\wwwroot</td></tr><tr class="alt"><th>Logon Method</th><td>&nbsp;&nbsp;&nbsp;Anonymous</td></tr><tr><th>Logon User</th><td>&nbsp;&nbsp;&nbsp;Anonymous</td></tr>
</table><div class="clear"></div></div></fieldset></div>
<div class="content-container"><fieldset><h4>More Information:</h4>This error occurs when a CGI application does not return a valid set of HTTP headers, or when a proxy or gateway was unable to send the request to a parent gateway. You may need to get a network trace or contact the proxy server administrator, if it is not a CGI problem.<p><a href="http://go.microsoft.com/fwlink/?LinkID=62293&amp;IIS70Error=502,0,0x00000000,14393">View more information &raquo;</a></p><p>Microsoft Knowledge Base Articles:</p>
</fieldset></div></div></body></html>
```

**Reply by Tratcher** [reply]
Turn all of the application logs in the proxy up to Debug and try again. Something more should be reported for a 502.
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0#configure-logging

**Reply by mikeesouth** [reply]
I'm finally getting back to this problem and I've enabled server logging. This is the server logs from YARP. I've left my full domain here (https://support.kattalo.com/) if there is some problem with the SSL certificate.
```
2021-03-03 13:10:40.077 +00:00 [Information] Microsoft.AspNetCore.Hosting.Diagnostics: Request starting HTTP/1.1 GET https://support.kattalo.com/ - -
2021-03-03 13:10:40.078 +00:00 [Information] Microsoft.AspNetCore.Routing.EndpointMiddleware: Executing endpoint 'route1'
2021-03-03 13:10:40.079 +00:00 [Information] Microsoft.ReverseProxy.Service.Proxy.HttpProxy: Proxying to https://custom.intercom.help/
2021-03-03 13:10:40.367 +00:00 [Information] Microsoft.ReverseProxy.Service.Proxy.HttpProxy: Request: An error was encountered before receiving a response.System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure: RemoteCertificateNameMismatchat System.Net.Security.SslStream.SendAuthResetSignal(ProtocolToken message, ExceptionDispatchInfo exception)at System.Net.Security.SslStream.ForceAuthenticationAsync[TIOAdapter](TIOAdapter adapter, Boolean receiveFirst, Byte[] reAuthenticationData, Boolean isApm)at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Boolean async, Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)--- End of inner exception stack trace ---at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Boolean async, Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)at System.Net.Http.HttpConnectionPool.GetHttp2ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)at Microsoft.ReverseProxy.Service.Proxy.HttpProxy.ProxyAsync(HttpContext context, String destinationPrefix, HttpMessageInvoker httpClient, RequestProxyOptions requestOptions, HttpTransformer transformer)
2021-03-03 13:10:40.368 +00:00 [Information] Microsoft.AspNetCore.Routing.EndpointMiddleware: Executed endpoint 'route1'
2021-03-03 13:10:40.375 +00:00 [Information] Microsoft.AspNetCore.Hosting.Diagnostics: Request finished HTTP/1.1 GET https://support.kattalo.com/ - - - 502 - - 297.2714ms
```

From this response I see that an option could be to set the Subject Alternative Name, is that possible in YARP? https://stackoverflow.com/questions/60341743/why-do-i-get-remotecertificatenamemismatch

**Reply by Tratcher** [reply]
`RemoteCertificateNameMismatch` This is pretty common with proxies. What is the current name in the destination's certificate? The request Host header needs to match that, or you need to override the cert validation.

YARP by default removes the Host header from incoming requests and uses the one given in the Destination Address field (`custom.intercom.help` above). You've set RequestHeaderOriginalHost=true to disable this, so the Host header is the one originally sent by the client. If you need some other value entirely then you can set it like this:
https://microsoft.github.io/reverse-proxy/articles/transforms.html#requestheader
```
{
  "RequestHeader": "Host",
  "Set": "MyCustomHost",
}
```

**Reply by mikeesouth** [reply]
> `RemoteCertificateNameMismatch` This is pretty common with proxies. What is the current name in the destination's certificate? The request Host header needs to match that, or you need to override the cert validation.
> 
> YARP by default removes the Host header from incoming requests and uses the one given in the Destination Address field (`custom.intercom.help` above). You've set RequestHeaderOriginalHost=true to disable this, so the Host header is the one originally sent by the client. If you need some other value entirely then you can set it like this:
> https://microsoft.github.io/reverse-proxy/articles/transforms.html#requestheader
> 
> ```
> {
>   "RequestHeader": "Host",
>   "Set": "MyCustomHost",
> }
> ```

Cheers! This fixed my problem. The expected domain was just "intercom.help", not "custom.intercom.help".
YARP is doing exactly what I want with minimal config. This is my final config:
```
  "ReverseProxy": {
    "Routes": [
      {
        "RouteId": "route1",
        "ClusterId": "cluster1",
        "Match": {
          "Path": "{**catch-all}"
        },
        "Transforms": [
          {
            "RequestHeader": "Host",
            "Set": "intercom.help"
          }
        ]
      }
    ],
    "Clusters": {
      "cluster1": {
        "Destinations": {
          "cluster1/destination1": {
            "Address": "https://custom.intercom.help/"
          }
        }
      }
    }
  }
```

I hope this helps someone else looking for intercom custom domain support.

**Comment by samsp-msft** [comment]
I would suggest you try using cURL or postman to figure out how to craft requests that are successful with the destination server, and then map those back to YARP once you understand exactly what is needed. It may be the config for certs etc on the  destination server that needs tweaking.

**Reply by mikeesouth** [reply]
Thanks for the tip. I'm familiar with postman to test an API but not enough to test these calls unfortunately.
